### PR TITLE
add notes on enabling strict dependency management

### PR DESCRIPTION
**Pull request by SamRoberts**

### CODE REVIEW

File: doc/strict_dependency_management.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
+# Strict conflict management
+
+Uniform-dependency contains a `strictDependencySettings` setting which you can add to your build,
+causing Ivy to use strict conflict management, resulting in build failures if you depend on
+conflicting versions of any library. This document describes methods to debug and fix conflicts.
+
+## Terminology, briefly
+
+I may use the words "library" or "module" to refer to a single body of code packaged together,
```
Comment (vineethvarghese) [inline doc/strict_dependency_management.md]: Do we want documentation written in first person?

Comment (SamRoberts) [inline doc/strict_dependency_management.md]: Good point. I find first person easier to read, personally. But I know it's not standard. I'll change it.

File: doc/strict_dependency_management.md [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+In your sbt file you will specify a list of dependencies. When you ask sbt to find the transitive
+forest of all the modules you depend on, sbt:
+
+  1. Creates some Ivy files describing your configurations and dependency list. These files are
+     located in the `target/resolution-cache/<organization>/<name>/<version-specific-path>/`
+     directory.
+  2. Asks Ivy to resolve your dependencies.
+  3. Ivy finds the transitive list of all modules you transitively depend on.
+  4. Ivy resolves conflicting versions these modules, using the specified conflict manager.
+     The default conflict manager picks the version it guesses is the "latest".
```
Comment (vineethvarghese) [inline doc/strict_dependency_management.md]: I presume the convention ivy uses to guess the latest is the x.x.x format?

File: doc/strict_dependency_management.md [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+  5. Ivy creates files listing all the modules you transitvely depend on, and the dependency
+     information between them. There is one of these files for each Ivy configuration, located
```
Comment (vineethvarghese) [inline doc/strict_dependency_management.md]: I know this is not an ivy documentation and I am not sure to what level of detail you intend to cover but what is Ivy configuration in `There is one of these files for each Ivy configuration`?

File: doc/strict_dependency_management.md [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+      </revision>
+    </module>
+
+Each `module` contains a list of `revision`s. Evicted revisions are marked as such. Each revision contains
+a list of the modules which depend on that revision. You can trace callers back from the conflicting
+module to build up a dependency graph, culminating in the dependencies which introduce the conflicts.
+
+Some notes:
+
+  -  When a module `A` depends on `B-v1`, which gets evicted in favour of `B-v2`, Ivy keeps the original
```
Comment (stephanh) [inline doc/strict_dependency_management.md]: Is it possible to get an example for this?

File: doc/strict_dependency_management.md (hunk 2 above)
Comment (SamRoberts) [inline doc/strict_dependency_management.md]: More or less. Covers a wider range of separators, adds separators between unseparated alphabetic and numeric sub-strings, handles some special strings (think "dev", "rc", etc).

Lots of magic: http://grepcode.com/file/repo1.maven.org/maven2/org.apache.ivy/ivy/2.4.0/org/apache/ivy/plugins/latest/LatestRevisionStrategy.java

File: doc/strict_dependency_management.md (hunk 3 above)
Comment (SamRoberts) [inline doc/strict_dependency_management.md]: Hrmm. Might add some links to Ivy documentation?

Comment (stephanh) [inline doc/strict_dependency_management.md]: Good idea.

**Comment by stephanh** [comment]
Can you also add a link to the doc here https://github.com/CommBank/uniform/blob/master/src/site/index.md.

**Comment by stephanh** [comment]
:+1: from me pending Vineeth's comments.

File: doc/strict_dependency_management.md [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+### Exclude module from dependencies with bad versions
+
+You can exclude a module from the transitive graph of modules pulled in by a particular dependency. For
+example, if you were to include both specs2 and scalacheck, you might need to exclude scalacheck from specs2:
+
+    "org.scalacheck" %% "scalacheck" % depend.versions.scalacheck,
+    "org.specs2"     %% "specs2"     % depend.versions.specs exclude("org.scalacheck", "scalacheck_2.10")
+
+This is not modular: the two depencies are no longer independant, but it has the advantage that that the
```
Comment (vineethvarghese) [inline doc/strict_dependency_management.md]: depencies -> dependencies

File: doc/strict_dependency_management.md [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+### Dependency overrides
+
+Inside sbt, we sometimes use dependency overrides to pin the version of a conflicting module:
+
+    // conflict between hive-exec / sqoop vs. avro-mapred
+    dependencyOverrides += "org.apache.velocity" % "velocity"  % "1.7",
+
+Dependency overrides are a poor way to resolve conflicts. The conflicting module will stay pinned at the
+specified version, even if the depencies which introduce the conflict switch to different versions.
```
Comment (vineethvarghese) [inline doc/strict_dependency_management.md]: depencies -> dependencies

File: doc/strict_dependency_management.md (hunk 4 above)
Comment (SamRoberts) [inline doc/strict_dependency_management.md]: Sure

File: doc/strict_dependency_management.md (hunk 5 above)
Comment (SamRoberts) [inline doc/strict_dependency_management.md]: thank you

File: doc/strict_dependency_management.md (hunk 6 above)
Comment (SamRoberts) [inline doc/strict_dependency_management.md]: thanks

**Comment by vineethvarghese** [comment]
Other than some minor comments, it looks good :+1:

**Pull request merged**